### PR TITLE
Clean asynchronous stop_callback implementation

### PR DESCRIPTION
**Pull request by gpotter2**
This PR:
- adds an asynchronous way of stopping `sniff`
- **WILL BREAK started_callback usage. However, it only exists in master (was not on 2.4.0), so we can allow such a thing**
- fixes https://github.com/secdev/scapy/issues/1505

image

I will add the unit tests if we agree on the PR

### CODE REVIEW

File: scapy/sendrecv.py [hunk 1]
```diff
@@ -897,7 +903,7 @@ def _select(sockets):
         read_allowed_exceptions = (PcapTimeoutElapsed,)
 
         def _select(sockets):
-            return sockets
+            return select_objects(sockets, remain=0)
```
Comment (gpotter2) [inline scapy/sendrecv.py]: Note to self: separate
- `WINDOWS and started_callback` ==> use select_objects
- `WINDOWS and not started_callback` ==> previous behavior: dont use select_objects

**Comment by guedou** [comment]
@gpotter2 what is the status of this PR?

**Comment by voi98306** [comment]
This is a VERY useful feature, implemented in a way consistent with how it has been done in C since at least the mid 90's, and thus well battle proven. I would REALLY like to see this coming out in a release.

The only comment I have is that if you return something more resembling an object, rather than a mere stop_... function, you can easily in the future add additional reasons for communicating with the sniff function without breaking backwards compatibility.

So instead of sending the "stop_callback" function to "started_callback", send a communication structure allowing the user to call "commstruct.stop_sniffing()". That way, in the future you can add other methods, like "commstruct.extend_timeout()" or whatever. Other changes needed to accomodate new features (such as not merely clearing the select list when objpipe is written to) will be well hidden behind the interface.

I am a huge fan and really appreciate your work. I hope to see this feature in a not too distant release.

**Comment by gpotter2** [comment]
> The only comment I have is that if you return something more resembling an object, rather than a mere stop_... function, you can easily in the future add additional reasons for communicating with the sniff function without breaking backwards compatibility.

Good point. I think the best option would be to implement a `sniff_async` function (or a `threaded` argument to `sniff`) that would pop a sniff in a thread and return such an object.

**Comment by gpotter2** [comment]
@voi98306 It's not ready yet (some tests still fail), but there's a working demo over https://github.com/secdev/scapy/issues/1999

**Pull request closed without merging**